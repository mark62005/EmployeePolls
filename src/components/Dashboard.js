import PropTypes from "prop-types";
import { connect } from "react-redux";

const Dashboard = () => {
    return (
        <div>
            Dashboard
        </div>
    );
};

Dashboard.propTypes = {

};

export default connect()(Dashboard);