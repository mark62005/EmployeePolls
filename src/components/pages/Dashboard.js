import PropTypes from "prop-types";
import { useState } from "react";
import { connect } from "react-redux";
import { sortQuestionsIdByTimestamp } from "../../utils/helpers";
import QuestionList from "../QuestionList";
import Container from "react-bootstrap/Container";
import Tabs from "react-bootstrap/Tabs";
import Tab from "react-bootstrap/Tab";

const Dashboard = ({ allQuestionIds, newQuestionIds, doneQuestionIds }) => {
    const [ tabKey, setTabKey ] = useState("new");

    const CATEGORIES = [ "new", "done", "all" ];

    const handleSelect = (eventKey) => {
        setTabKey(eventKey);
    };

    const getListTitle = (category) => {
        switch (category) {
            case "new":
                return "Unanswered Questions";
            case "done":
                return "Answered Questions";
            case "all":
                return "All";
            default:
                return "";
        }
    };

    const renderQuestionList = (category) => {
        switch (category) {
            case "new":
                return <QuestionList
                    key={ category }
                    category={ category }
                    questionIds={ newQuestionIds }
                />;
            case "done":
                return <QuestionList
                    key={ category }
                    category={ category }
                    questionIds={ doneQuestionIds }
                />;
            case "all":
                return <QuestionList
                    key={ category }
                    category={ category }
                    questionIds={ allQuestionIds }
                />;
            default:
                return null;
        }
    };

    return (
        <Container className="p-3">
            <Tabs
                activeKey={ tabKey }
                onSelect={ handleSelect }
            >
                {
                    CATEGORIES.map((category) => (
                        <Tab
                            key={ category }
                            eventKey={ category }
                            title={ getListTitle(category) }
                        >
                            { renderQuestionList(category) }
                        </Tab>
                    ))
                }
            </Tabs>
        </Container>
    );
};

Dashboard.propTypes = {
    allQuestionIds: PropTypes.array.isRequired,
    newQuestionIds: PropTypes.array.isRequired,
    doneQuestionIds: PropTypes.array.isRequired,
};

function getFilteredQuestions(questions, authedUser, category) {
    switch (category) {
        case "new":
            return Object.values(questions).filter(
                (q) => !q.optionOne.votes.includes(authedUser) && !q.optionTwo.votes.includes(authedUser)
            );
        case "done":
            return Object.values(questions).filter(
                (q) => q.optionOne.votes.includes(authedUser) || q.optionTwo.votes.includes(authedUser)
            );
        case "all":
            return Object.values(questions);
        default:
            return [];
    }
}

const mapStateToProps = ({ authedUser, questions }) => {
    const newQuestions = getFilteredQuestions(questions, authedUser, "new");
    const doneQuestions = getFilteredQuestions(questions, authedUser, "done");

    return {
        authedUser,
        allQuestionIds: sortQuestionsIdByTimestamp(Object.values(questions)),
        newQuestionIds: sortQuestionsIdByTimestamp(newQuestions),
        doneQuestionIds: sortQuestionsIdByTimestamp(doneQuestions),
    };
};

export default connect(mapStateToProps)(Dashboard);