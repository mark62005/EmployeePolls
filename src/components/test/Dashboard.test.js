import { render } from "@testing-library/react";
import thunk from "redux-thunk";
import reducer from "../../reducers";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Dashboard from "../Dashboard";

describe("Dashboard", () => {
    var mockStore = configureStore({ reducer, middleware: [ thunk ] });

    it("will match the snapshot", () => {
        var view = render(
            <MemoryRouter>
                <Provider store={ mockStore }>
                    <Dashboard />
                </Provider>
            </MemoryRouter>
        );

        expect(view).toMatchSnapshot();
    });
});