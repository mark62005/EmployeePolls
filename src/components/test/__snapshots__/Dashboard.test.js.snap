// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard will match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="p-3 container"
      >
        <ul
          class="nav nav-tabs"
          role="tablist"
        >
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="react-aria4716161682-1-tabpane-new"
              aria-selected="true"
              class="nav-link active"
              data-rr-ui-event-key="new"
              id="react-aria4716161682-1-tab-new"
              role="tab"
              type="button"
            >
              Unanswered Questions
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="react-aria4716161682-1-tabpane-done"
              aria-selected="false"
              class="nav-link"
              data-rr-ui-event-key="done"
              id="react-aria4716161682-1-tab-done"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Answered Questions
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="react-aria4716161682-1-tabpane-all"
              aria-selected="false"
              class="nav-link"
              data-rr-ui-event-key="all"
              id="react-aria4716161682-1-tab-all"
              role="tab"
              tabindex="-1"
              type="button"
            >
              All
            </button>
          </li>
        </ul>
        <div
          class="tab-content"
        >
          <div
            aria-labelledby="react-aria4716161682-1-tab-new"
            class="fade tab-pane active show"
            id="react-aria4716161682-1-tabpane-new"
            role="tabpanel"
          >
            <div
              class="container"
            >
              <div
                class="border border-1 border-top-0 px-3 py-4 row"
              >
                <div
                  class="col"
                >
                  <div
                    class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="react-aria4716161682-1-tab-done"
            class="fade tab-pane"
            id="react-aria4716161682-1-tabpane-done"
            role="tabpanel"
          >
            <div
              class="container"
            >
              <div
                class="border border-1 border-top-0 px-3 py-4 row"
              >
                <div
                  class="col"
                >
                  <div
                    class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="react-aria4716161682-1-tab-all"
            class="fade tab-pane"
            id="react-aria4716161682-1-tabpane-all"
            role="tabpanel"
          >
            <div
              class="container"
            >
              <div
                class="border border-1 border-top-0 px-3 py-4 row"
              >
                <div
                  class="col"
                >
                  <div
                    class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="p-3 container"
    >
      <ul
        class="nav nav-tabs"
        role="tablist"
      >
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="react-aria4716161682-1-tabpane-new"
            aria-selected="true"
            class="nav-link active"
            data-rr-ui-event-key="new"
            id="react-aria4716161682-1-tab-new"
            role="tab"
            type="button"
          >
            Unanswered Questions
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="react-aria4716161682-1-tabpane-done"
            aria-selected="false"
            class="nav-link"
            data-rr-ui-event-key="done"
            id="react-aria4716161682-1-tab-done"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Answered Questions
          </button>
        </li>
        <li
          class="nav-item"
          role="presentation"
        >
          <button
            aria-controls="react-aria4716161682-1-tabpane-all"
            aria-selected="false"
            class="nav-link"
            data-rr-ui-event-key="all"
            id="react-aria4716161682-1-tab-all"
            role="tab"
            tabindex="-1"
            type="button"
          >
            All
          </button>
        </li>
      </ul>
      <div
        class="tab-content"
      >
        <div
          aria-labelledby="react-aria4716161682-1-tab-new"
          class="fade tab-pane active show"
          id="react-aria4716161682-1-tabpane-new"
          role="tabpanel"
        >
          <div
            class="container"
          >
            <div
              class="border border-1 border-top-0 px-3 py-4 row"
            >
              <div
                class="col"
              >
                <div
                  class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="react-aria4716161682-1-tab-done"
          class="fade tab-pane"
          id="react-aria4716161682-1-tabpane-done"
          role="tabpanel"
        >
          <div
            class="container"
          >
            <div
              class="border border-1 border-top-0 px-3 py-4 row"
            >
              <div
                class="col"
              >
                <div
                  class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="react-aria4716161682-1-tab-all"
          class="fade tab-pane"
          id="react-aria4716161682-1-tabpane-all"
          role="tabpanel"
        >
          <div
            class="container"
          >
            <div
              class="border border-1 border-top-0 px-3 py-4 row"
            >
              <div
                class="col"
              >
                <div
                  class="g-4 p-3 row row-cols-lg-3 row-cols-md-2"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
